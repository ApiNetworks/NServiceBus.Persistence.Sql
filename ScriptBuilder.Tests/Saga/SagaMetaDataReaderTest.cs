using System.IO;
using Mono.Cecil;
using NServiceBus;
using NServiceBus.Persistence.Sql;
using NUnit.Framework;
using ObjectApproval;

[TestFixture]
public class SagaMetaDataReaderTest
{
    ModuleDefinition module;

    public SagaMetaDataReaderTest()
    {
        var path = Path.Combine(TestContext.CurrentContext.TestDirectory, "ScriptBuilder.Tests.dll");
        module = ModuleDefinition.ReadModule(path);
    }

    [Test]
    public void Simple()
    {
        var dataType = module.GetTypeDefinition<SimpleSaga>();
        SagaDefinition definition;
        SagaDefinitionReader.TryGetSqlSagaDefinition(dataType, out definition);
        ObjectApprover.VerifyWithJson(definition);
    }

    [SqlSaga(
         correlationProperty: nameof(SagaData.Correlation),
         transitionalCorrelationProperty: nameof(SagaData.Transitional)
     )]
    public class SimpleSaga : Saga<SimpleSaga.SagaData>
    {
        public class SagaData : ContainSagaData
        {
            public string Correlation { get; set; }
            public string Transitional { get; set; }
        }

        protected override void ConfigureHowToFindSaga(SagaPropertyMapper<SagaData> mapper)
        {
        }
    }

    [Test]
    public void WithNoCorrelation()
    {
        var dataType = module.GetTypeDefinition<WithNoCorrelationSaga>();
        SagaDefinition definition;
        SagaDefinitionReader.TryGetSqlSagaDefinition(dataType, out definition);
        ObjectApprover.VerifyWithJson(definition);
    }

    [SqlSaga]
    public class WithNoCorrelationSaga : Saga<WithNoCorrelationSaga.SagaData>
    {
        public class SagaData : ContainSagaData
        {
        }

        protected override void ConfigureHowToFindSaga(SagaPropertyMapper<SagaData> mapper)
        {
        }
    }

    [Test]
    public void WithNoTransitionalCorrelation()
    {
        var dataType = module.GetTypeDefinition<WithNoTransitionalCorrelationSaga>();
        SagaDefinition definition;
        SagaDefinitionReader.TryGetSqlSagaDefinition(dataType, out definition);
        ObjectApprover.VerifyWithJson(definition);
    }

    [SqlSaga]
    public class WithNoTransitionalCorrelationSaga : Saga<WithNoTransitionalCorrelationSaga.SagaData>
    {
        public class SagaData : ContainSagaData
        {
        }

        protected override void ConfigureHowToFindSaga(SagaPropertyMapper<SagaData> mapper)
        {
        }
    }

}