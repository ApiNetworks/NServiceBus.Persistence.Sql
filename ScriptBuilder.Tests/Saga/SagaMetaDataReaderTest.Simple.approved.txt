{
  "TableSuffix": "SimpleSaga",
  "CorrelationProperty": {
    "Name": "Correlation",
    "Type": "String"
  },
  "TransitionalCorrelationProperty": {
    "Name": "Transitional",
    "Type": "String"
  },
  "Name": "SagaMetaDataReaderTest/SimpleSaga"
}